### PR TITLE
General update and fix of CPU/GPU bug arising from latest multi-threaded Caffe

### PR DESCRIPTION
**Pull request by GiuliaP**
This PR
1. Fixes issue https://github.com/robotology/himrep/issues/18 for which the feature extraction happens to be in CPU mode (default) unless the GPU mode (and device ID) are explicitly set in the same thread performing the extraction, since in newest `caffe` these settings are thread-specific (thanks @Tobias-Fischer ). **This change cannot break backward compatibility.**
2. Adopts `float` precision instead of `double` for feature extraction (which is `2x` faster and accurate enough), but leaving the same output format, so that we do not have to change other modules: the feature is extracted as `float` by `caffe` and then converted to `double` in order to be sent as output through the `/caffeCoder/code:o` port or dumped to binary file. **This change cannot break backward compatibility.**
3. Replaces the `ForwardPrefilled()` API call with `Forward()` since the previous one is deprecated in newest `caffe`. I haven't checked the compatibility of the `Forward()` call with older `caffe`: indeed with older `caffe` I was using the deprecated call since `Forward()` was not providing the desired streaming behavior, therefore  **this may break backward compatibility (by producing, for instance, as output, the representation of the same image, repeatedly).**
4. updates `.prototxt` files according to latest `caffe` and **this change may break backward compatibility (by producing a `caffe` runtime error).**

I have checked that, with newest `caffe`, the behavior is unchanged with these four updates.

Instead, I wrote **cannot break** or **may break** backward compatibility because I haven't tested this code with old `caffe` versions (@pattacini for instance the commit currently installed on the tesla machine of the black robot). Indeed, we agreed that it is impossible to verify against every older `caffe` commit before the introduction of the versioning system by `caffe` maintainers. On the contrary, from now on we will guarantee to be compatible with specific versions starting from the current one.
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/robotology/himrep/pull/19%23issuecomment-220003634%22%2C%20%22https%3A//github.com/robotology/himrep/pull/19%23issuecomment-220009773%22%2C%20%22https%3A//github.com/robotology/himrep/pull/19%23discussion_r64197074%22%2C%20%22https%3A//github.com/robotology/himrep/pull/19%23issuecomment-220939591%22%2C%20%22https%3A//github.com/robotology/himrep/pull/19%23issuecomment-221274727%22%2C%20%22https%3A//github.com/robotology/himrep/pull/19%23issuecomment-222448587%22%2C%20%22https%3A//github.com/robotology/himrep/pull/19%23discussion_r65046329%22%2C%20%22https%3A//github.com/robotology/himrep/pull/19%23issuecomment-222451011%22%2C%20%22https%3A//github.com/robotology/himrep/pull/19%23issuecomment-223812678%22%2C%20%22https%3A//github.com/robotology/himrep/pull/19%23issuecomment-223878383%22%2C%20%22https%3A//github.com/robotology/himrep/pull/19%23issuecomment-226500114%22%2C%20%22https%3A//github.com/robotology/himrep/pull/19%23issuecomment-226774215%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/robotology/himrep/pull/19%23issuecomment-220003634%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Thanks%20a%20lot%20%40GiuliaP%20for%20this%20PR.%5Cr%5Cn%5Cr%5CnHere%27s%20a%20tentative%20schedule%20on%20how%20to%20proceed%3A%5Cr%5Cn-%20%60Blacky%60%20is%20leaving%20on%20mission%20next%20week.%20We%20can%20wait%20until%20he%27s%20back%20and%20go%20on%20with%20the%20upgrade.%5Cr%5Cn-%20In%20two%20weeks%20%40Tobias-Fischer%20will%20be%20also%20back%20from%20ICRA%20and%20could%20test%20this%20PR.%22%2C%20%22created_at%22%3A%20%222016-05-18T11%3A52%3A40Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3738070%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pattacini%22%7D%7D%2C%20%7B%22body%22%3A%20%22Ok%21%22%2C%20%22created_at%22%3A%20%222016-05-18T12%3A22%3A25Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6449556%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/GiuliaP%22%7D%7D%2C%20%7B%22body%22%3A%20%22Hey%2C%20indeed%20I%27m%20back%20from%20ICRA%20and%20we%27ll%20have%20a%20look%20at%20this%20today%20or%20tomorrow.%20Thanks%20a%20lot%20for%20the%20fix%20%3A%29%21%22%2C%20%22created_at%22%3A%20%222016-05-23T10%3A07%3A00Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/5497832%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/Tobias-Fischer%22%7D%7D%2C%20%7B%22body%22%3A%20%22Hi%2C%5Cr%5Cnit%27s%20working%20perfectly%20here%20in%20London.%20Fine%20to%20merge%20from%20my%20side%2C%20I%20guess%20you%20guys%20can%20merge%20as%20soon%20as%20you%20tested%20it%20on%20your%20iCub%20%3A%29.%5Cr%5Cn%5Cr%5CnThanks%20a%20lot%2C%5Cr%5CnTobi%22%2C%20%22created_at%22%3A%20%222016-05-24T13%3A47%3A27Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/5497832%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/Tobias-Fischer%22%7D%7D%2C%20%7B%22body%22%3A%20%22Hi%20%40GiuliaP%20%5Cr%5Cn%5Cr%5CnOn%20the%207th%20we%27ll%20have%20a%20fairly%20important%20demo%2C%20so%20I%20think%20we%20can%20postpone%20the%20upgrade%20after%20that%20date%2C%20to%20avoid%20any%20possible%20glitch.%22%2C%20%22created_at%22%3A%20%222016-05-30T09%3A09%3A59Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3738070%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pattacini%22%7D%7D%2C%20%7B%22body%22%3A%20%22Hi%2C%20ok%21%20%22%2C%20%22created_at%22%3A%20%222016-05-30T09%3A22%3A19Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6449556%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/GiuliaP%22%7D%7D%2C%20%7B%22body%22%3A%20%22Hi%20%40GiuliaP%20%5Cr%5Cn%5Cr%5CnThere%27s%20a%20major%20overhaul%20of%20the%20Blacky%27s%20system%20scheduled%20by%20%40mbrunettini%20on%20the%20%2A%2A13th%20of%20June%2A%2A%2C%20If%20I%20remember%20correctly.%5Cr%5Cn%5Cr%5CnCould%20you%20coordinate%20with%20him%20for%20proceeding%20with%20Caffe%20upgrade%3F%5Cr%5CnUnfortunately%2C%20the%20system%20cannot%20be%20touched%20until%20the%2013th%2C%20given%20that%20we%27ll%20have%20yet%20another%20mission%20on%2010/06%20%3Adisappointed%3A.%22%2C%20%22created_at%22%3A%20%222016-06-05T13%3A18%3A50Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3738070%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pattacini%22%7D%7D%2C%20%7B%22body%22%3A%20%22Hi%20%40pattacini%20%2C%20ok%21%20We%20will%20coordinate%20in%20order%20to%20upgrade%20the%20Tesla%20machine%20accordingly%2C%20updating%20the%20instructions%20and%20all%20Caffe-related%20software%20in%20order%20to%20run%20on%20Ubuntu%2016.04%20%28we%20already%20talked%20about%20this%20with%20%40mbrunettini%29.%22%2C%20%22created_at%22%3A%20%222016-06-06T06%3A31%3A19Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6449556%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/GiuliaP%22%7D%7D%2C%20%7B%22body%22%3A%20%22I%20have%20updated%20the%20instructions%2C%20by%20reporting%20what%20we%20did%20with%20%40mbrunettini%20to%20install%20Caffe%20on%20Ubuntu%2016.04%20on%20our%20Tesla%20k40.%20We%20could%20wait%20that%20%40mbrunettini%20check%20them%2C%20while%20upgrading%20also%20the%20second%20machine%20that%20we%20have%2C%20and%20merge%20the%20PR%20afterwards%2C%20given%20that%20the%20code%20seems%20to%20work%20%28%40Tobias-Fischer%20checked%20it%20and%20I%20also%20have%20checked%20it%20with%20%40barbalberto%20on%20the%20Tegra%20card%20that%20we%20have%29.%22%2C%20%22created_at%22%3A%20%222016-06-16T14%3A22%3A33Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6449556%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/GiuliaP%22%7D%7D%2C%20%7B%22body%22%3A%20%22Hi%20%40GiuliaP%2C%5Cr%5CnI%20can%20confirm%20that%20your%20changed%20instructions%20are%20okay.%20I%27ve%20recently%20installed%20%60himrep%60%20on%20Ubuntu%2016.04%20as%20well%2C%20and%20the%20changes%20you%20describe%20are%20exactly%20the%20ones%20I%20did.%5Cr%5CnI%27d%20be%20happy%20to%20see%20this%20PR%20merged%20sooner%20rather%20than%20later%20%3A%29.%5Cr%5Cn%5Cr%5CnBest%2C%5Cr%5CnTobi%22%2C%20%22created_at%22%3A%20%222016-06-17T13%3A52%3A16Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/5497832%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/Tobias-Fischer%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%20f1f928adfabe5105a14dbe3d5ae8deec005901f9%20modules/caffeCoder/src/CaffeFeatExtractor.hpp%2010%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/robotology/himrep/pull/19%23discussion_r64197074%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22I%20think%20we%20can%20get%20rid%20of%20this%2C%20as%20we%20decided%20to%20require%20Caffe%20version%201%20or%20bigger.%22%2C%20%22created_at%22%3A%20%222016-05-23T10%3A01%3A59Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/5497832%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/Tobias-Fischer%22%7D%7D%2C%20%7B%22body%22%3A%20%22True%3A%20I%20think%20we%20can%20leave%20it%20as%20it%20is%20the%20same%20kind%20of%20check%20that%20we%20will%20need%20to%20do%20in%20the%20near%20future%20on%20the%20different%20versions%2C%20but%20we%20can%20modify%20it%20for%20the%20moment%20in%20order%20to%20just%20require%20version%20higher%20than%201.%22%2C%20%22created_at%22%3A%20%222016-05-30T09%3A22%3A13Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6449556%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/GiuliaP%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20modules/caffeCoder/src/CaffeFeatExtractor.hpp%3AL19-36%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'></a>
- [ ] <a href='#crh-comment-General Comment'></a> &nbsp;<b><a href='https://github.com/robotology/himrep/pull/19#issuecomment-220003634'>General Comment</a></b>
- <a href='https://github.com/pattacini'></a> Thanks a lot @GiuliaP for this PR.
  Here's a tentative schedule on how to proceed:
- `Blacky` is leaving on mission next week. We can wait until he's back and go on with the upgrade.
- In two weeks @Tobias-Fischer will be also back from ICRA and could test this PR.
- <a href='https://github.com/GiuliaP'></a> Ok!
- <a href='https://github.com/Tobias-Fischer'></a> Hey, indeed I'm back from ICRA and we'll have a look at this today or tomorrow. Thanks a lot for the fix :)!
- <a href='https://github.com/Tobias-Fischer'></a> Hi,
  it's working perfectly here in London. Fine to merge from my side, I guess you guys can merge as soon as you tested it on your iCub :).
  Thanks a lot,
  Tobi
- <a href='https://github.com/pattacini'></a> Hi @GiuliaP
  On the 7th we'll have a fairly important demo, so I think we can postpone the upgrade after that date, to avoid any possible glitch.
- <a href='https://github.com/GiuliaP'></a> Hi, ok!
- <a href='https://github.com/pattacini'></a> Hi @GiuliaP
  There's a major overhaul of the Blacky's system scheduled by @mbrunettini on the **13th of June**, If I remember correctly.
  Could you coordinate with him for proceeding with Caffe upgrade?
  Unfortunately, the system cannot be touched until the 13th, given that we'll have yet another mission on 10/06 :disappointed:.
- <a href='https://github.com/GiuliaP'></a> Hi @pattacini , ok! We will coordinate in order to upgrade the Tesla machine accordingly, updating the instructions and all Caffe-related software in order to run on Ubuntu 16.04 (we already talked about this with @mbrunettini).
- <a href='https://github.com/GiuliaP'></a> I have updated the instructions, by reporting what we did with @mbrunettini to install Caffe on Ubuntu 16.04 on our Tesla k40. We could wait that @mbrunettini check them, while upgrading also the second machine that we have, and merge the PR afterwards, given that the code seems to work (@Tobias-Fischer checked it and I also have checked it with @barbalberto on the Tegra card that we have).
- <a href='https://github.com/Tobias-Fischer'></a> Hi @GiuliaP,
  I can confirm that your changed instructions are okay. I've recently installed `himrep` on Ubuntu 16.04 as well, and the changes you describe are exactly the ones I did.
  I'd be happy to see this PR merged sooner rather than later :).
  Best,
  Tobi
- [ ] <a href='#crh-comment-Pull f1f928adfabe5105a14dbe3d5ae8deec005901f9 modules/caffeCoder/src/CaffeFeatExtractor.hpp 10'></a> &nbsp;<b><a href='https://github.com/robotology/himrep/pull/19#discussion_r64197074'>File: modules/caffeCoder/src/CaffeFeatExtractor.hpp:L19-36</a></b>
- <a href='https://github.com/Tobias-Fischer'></a> I think we can get rid of this, as we decided to require Caffe version 1 or bigger.
- <a href='https://github.com/GiuliaP'></a> True: I think we can leave it as it is the same kind of check that we will need to do in the near future on the different versions, but we can modify it for the moment in order to just require version higher than 1.

<a href='https://www.codereviewhub.com/robotology/himrep/pull/19?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/robotology/himrep/pull/19?approve=1'></a>&nbsp;<a href='https://github.com/robotology/himrep/pull/19'></a>
<a href='#crh-end'></a>

### CODE REVIEW

**Comment by pattacini** [comment]
This pull request is now managed with CodeReviewHub. <a href='https://github.com/robotology/himrep/pull/19'>Refresh</a>.<a href='#crh-update-none'></a>

**Comment by pattacini** [comment]
Thanks a lot @GiuliaP for this PR.

Here's a tentative schedule on how to proceed:
- `Blacky` is leaving on mission next week. We can wait until it's back and go on with the upgrade.
- In two weeks @Tobias-Fischer will be also back from ICRA and could test this PR.

**Comment by GiuliaP** [comment]
Ok!

File: modules/caffeCoder/src/CaffeFeatExtractor.hpp [hunk 1]
```diff
@@ -19,17 +19,17 @@
 #include "caffe-version.h"
 
 #if (CAFFE_MAJOR >= 1)
-	// new caffe headers
-	#include "caffe/caffe.hpp"
- 	#include "caffe/layers/memory_data_layer.hpp"
+    // new caffe headers
+    #include "caffe/caffe.hpp"
+    #include "caffe/layers/memory_data_layer.hpp"
 #else
```
Comment (Tobias-Fischer) [inline modules/caffeCoder/src/CaffeFeatExtractor.hpp:25]: I think we can get rid of this, as we decided to require Caffe version 1 or bigger.

**Comment by Tobias-Fischer** [comment]
Hey, indeed I'm back from ICRA and we'll have a look at this today or tomorrow. Thanks a lot for the fix :)!

**Comment by Tobias-Fischer** [comment]
Hi,
it's working perfectly here in London. Fine to merge from my side, I guess you guys can merge as soon as you tested it on your iCub :).

Thanks a lot,
Tobi

**Comment by pattacini** [comment]
Hi @GiuliaP 

On the 7th we'll have a fairly important demo, so I think we can postpone the upgrade after that date, to avoid any possible glitch.

File: modules/caffeCoder/src/CaffeFeatExtractor.hpp (hunk 1 above)
Comment (GiuliaP) [inline modules/caffeCoder/src/CaffeFeatExtractor.hpp:25]: True: I think we can leave it as it is the same kind of check that we will need to do in the near future on the different versions, but we can modify it for the moment in order to just require version higher than 1.

**Comment by GiuliaP** [comment]
Hi, ok!

**Comment by pattacini** [comment]
Hi @GiuliaP 

There's a major overhaul of the Blacky's system scheduled by @mbrunettini on the **13th of June**, If I remember correctly.

Could you coordinate with him for proceeding with Caffe upgrade?
Unfortunately, the system cannot be touched until the 13th, given that we'll have yet another mission on 10/06 :disappointed:.

**Comment by GiuliaP** [comment]
Hi @pattacini , ok! We will coordinate in order to upgrade the Tesla machine accordingly, updating the instructions and all Caffe-related software in order to run on Ubuntu 16.04 (we already talked about this with @mbrunettini).

**Comment by GiuliaP** [comment]
I have updated the instructions, by reporting what we did with @mbrunettini to install Caffe on Ubuntu 16.04 on our Tesla k40. We could wait that @mbrunettini check them, while upgrading also the second machine that we have, and merge the PR afterwards, given that the code seems to work (@Tobias-Fischer checked it and I also have checked it with @barbalberto on the Tegra card that we have).

**Comment by Tobias-Fischer** [comment]
Hi @GiuliaP,
I can confirm that your changed instructions are okay. I've recently installed `himrep` on Ubuntu 16.04 as well, and the changes you describe are exactly the ones I did.
I'd be happy to see this PR merged sooner rather than later :).

Best,
Tobi

**Pull request merged**